File: backend/internal/model/cicd.go
package model

import (
	"encoding/json"
	"time"
)

// CICDProvider identifies the CI/CD platform a config targets.
//
// Supported values map to PRD §14.1:
//
//	github_actions  — GitHub Actions (Webhook + API)
//	gitlab_ci       — GitLab CI (Webhook + API)
//	jenkins         — Jenkins (API)
//	generic         — any webhook-driven pipeline (default)
type CICDProvider string

const (
	CICDProviderGitHubActions CICDProvider = "github_actions"
	CICDProviderGitLabCI      CICDProvider = "gitlab_ci"
	CICDProviderJenkins       CICDProvider = "jenkins"
	CICDProviderGeneric       CICDProvider = "generic"
)

// CICDConfig captures the connection details for a CI/CD pipeline
// (PRD P4-003: cicd_configs table).
//
// A config is workspace-scoped and may optionally be project-scoped. It
// records the provider, API endpoint, secret token reference, and the events
// that should auto-trigger a build. The token is stored as an opaque
// reference (e.g. an env-var key or secret manager handle) rather than the
// raw credential — the service resolves it at trigger time.
type CICDConfig struct {
	BaseModel

	WorkspaceID uint64  `gorm:"not null;index" json:"workspace_id"`
	ProjectID   *uint64 `gorm:"index" json:"project_id,omitempty"`

	Name string `gorm:"size:120;not null" json:"name"`

	// Provider: github_actions | gitlab_ci | jenkins | generic
	Provider CICDProvider `gorm:"size:30;not null;default:generic" json:"provider"`

	// APIEndpoint is the provider's API base URL or webhook trigger URL.
	APIEndpoint string `gorm:"size:512" json:"api_endpoint"`

	// ProjectSlug is the provider-specific project/repo identifier
	// (e.g. "owner/repo" for GitHub, project path for GitLab, job name for Jenkins).
	ProjectSlug string `gorm:"size:255" json:"project_slug"`

	// DefaultBranch is the branch to build when none is supplied.
	DefaultBranch string `gorm:"size:120;default:main" json:"default_branch"`

	// AuthTokenRef references the stored credential. We never persist the raw
	// token — only an opaque reference (env var name, secret key, etc.).
	AuthTokenRef string `gorm:"size:255" json:"auth_token_ref,omitempty"`

	// TriggerEvents: array of event names that should auto-trigger a build
	// (e.g. ["push","pull_request","manual"]). Stored as JSONB.
	TriggerEvents json.RawMessage `gorm:"type:jsonb;default:'[]'" json:"trigger_events"`

	// ExtraConfig: provider-specific free-form config (JSONB).
	ExtraConfig json.RawMessage `gorm:"type:jsonb;default:'{}'" json:"extra_config"`

	Enabled bool `gorm:"default:true" json:"enabled"`

	// Relationships
	Workspace Workspace `gorm:"foreignKey:WorkspaceID" json:"-"`
	Project   *Project  `gorm:"foreignKey:ProjectID" json:"-"`
}

func (CICDConfig) TableName() string { return "cicd_configs" }

// BuildStatus defines the lifecycle state of a build (PRD P4-003 build_records).
//
// Flow: pending → queued → running → success | failed | cancelled
// Unknown is used when the provider returns an unrecognized state.
type BuildStatus string

const (
	BuildPending   BuildStatus = "pending"
	BuildQueued    BuildStatus = "queued"
	BuildRunning   BuildStatus = "running"
	BuildSuccess   BuildStatus = "success"
	BuildFailed    BuildStatus = "failed"
	BuildCancelled BuildStatus = "cancelled"
	BuildUnknown   BuildStatus = "unknown"
)

// BuildTrigger identifies what initiated a build.
type BuildTrigger string

const (
	BuildTriggerManual   BuildTrigger = "manual"
	BuildTriggerPush     BuildTrigger = "push"
	BuildTriggerPull     BuildTrigger = "pull_request"
	BuildTriggerSchedule BuildTrigger = "schedule"
	BuildTriggerAgent    BuildTrigger = "agent"
	BuildTriggerWebhook  BuildTrigger = "webhook"
)

// BuildRecord tracks a single CI/CD pipeline execution
// (PRD P4-003: build_records table).
//
// A build is created when a CICDConfig is triggered (manually or by an event).
// The service records the trigger source, branch, commit, provider build id,
// per-stage progress, and the final outcome. Status updates flow through SSE
// as cicd_build.* events so the UI can render live progress.
type BuildRecord struct {
	BaseModel

	WorkspaceID  uint64  `gorm:"not null;index" json:"workspace_id"`
	ProjectID    *uint64 `gorm:"index" json:"project_id,omitempty"`
	CICDConfigID uint64  `gorm:"not null;index" json:"cicd_config_id"`

	// Trigger: manual | push | pull_request | schedule | agent | webhook
	Trigger BuildTrigger `gorm:"size:20;not null;default:manual" json:"trigger"`

	// Source references
	Branch     string  `gorm:"size:120" json:"branch"`
	CommitSHA  string  `gorm:"size:64" json:"commit_sha"`
	IssueID    *uint64 `gorm:"index" json:"issue_id,omitempty"`
	AgentTaskID *uint64 `gorm:"index" json:"agent_task_id,omitempty"`
	TriggeredByID uint64 `gorm:"column:triggered_by_id" json:"triggered_by_id"`

	// Provider-side identifiers
	ExternalBuildID string `gorm:"size:128" json:"external_build_id"`
	BuildURL        string `gorm:"size:512" json:"build_url"`

	// Stages: array of {name, status, duration_ms, started_at, completed_at}
	Stages json.RawMessage `gorm:"type:jsonb;default:'[]'" json:"stages"`

	// Lifecycle
	Status       BuildStatus `gorm:"size:20;default:pending;index" json:"status"`
	Progress     int         `gorm:"default:0" json:"progress"`
	CurrentStage *string     `gorm:"size:100" json:"current_stage,omitempty"`
	ErrorMessage *string     `gorm:"type:text" json:"error_message,omitempty"`

	StartedAt   *time.Time `json:"started_at,omitempty"`
	CompletedAt *time.Time `json:"completed_at,omitempty"`
	CancelledAt *time.Time `json:"cancelled_at,omitempty"`

	// Roll-up
	DurationMs int64 `gorm:"default:0" json:"duration_ms"`

	// Relationships
	Workspace  Workspace  `gorm:"foreignKey:WorkspaceID" json:"-"`
	Project    *Project   `gorm:"foreignKey:ProjectID" json:"-"`
	CICDConfig CICDConfig `gorm:"foreignKey:CICDConfigID" json:"-"`
}

func (BuildRecord) TableName() string { return "build_records" }

File: frontend/src/api/cicd.ts
/**
 * CI/CD API - PRD P4-005
 *
 * Manage CI/CD configs (workspace- or project-scoped) and trigger/monitor
 * builds. Each build runs an async workflow whose progress streams in via
 * SSE events `cicd_build.*` (see useSSE.ts).
 */
import api from './index'

export type CICDProvider =
  | 'github_actions'
  | 'gitlab_ci'
  | 'jenkins'
  | 'generic'

export type BuildStatus =
  | 'pending'
  | 'queued'
  | 'running'
  | 'success'
  | 'failed'
  | 'cancelled'
  | 'unknown'

export type BuildTrigger =
  | 'manual'
  | 'push'
  | 'pull_request'
  | 'schedule'
  | 'agent'
  | 'webhook'

export interface BuildStage {
  name: string
  status: 'pending' | 'running' | 'success' | 'failed' | 'skipped'
  duration_ms: number
  started_at?: string
  completed_at?: string
  log_url?: string
}

export interface CICDConfig {
  id: number
  workspace_id: number
  project_id?: number
  name: string
  provider: CICDProvider
  api_endpoint: string
  project_slug: string
  default_branch: string
  auth_token_ref: string
  trigger_events: string[]
  extra_config: Record<string, unknown>
  enabled: boolean
  created_at: string
  updated_at: string
}

export interface CICDConfigCreate {
  name: string
  provider?: CICDProvider
  api_endpoint?: string
  project_slug?: string
  default_branch?: string
  auth_token_ref?: string
  trigger_events?: string[]
  extra_config?: Record<string, unknown>
  project_id?: number
  enabled?: boolean
}

export interface CICDConfigUpdate {
  name?: string
  provider?: CICDProvider
  api_endpoint?: string
  project_slug?: string
  default_branch?: string
  auth_token_ref?: string
  trigger_events?: string[]
  extra_config?: Record<string, unknown>
  enabled?: boolean
}

export interface CICDConfigListFilter {
  project_id?: number
}

export interface BuildRecord {
  id: number
  workspace_id: number
  project_id?: number
  cicd_config_id: number
  cicd_config_name?: string
  trigger: BuildTrigger
  branch: string
  commit_sha: string
  issue_id?: number
  agent_task_id?: number
  triggered_by_id: number
  external_build_id: string
  build_url: string
  stages: BuildStage[]
  status: BuildStatus
  progress: number
  current_stage?: string
  error_message?: string
  started_at?: string
  completed_at?: string
  cancelled_at?: string
  duration_ms: number
  created_at: string
  updated_at: string
}

export interface BuildTriggerRequest {
  cicd_config_id: number
  branch?: string
  commit_sha?: string
  trigger?: BuildTrigger
  project_id?: number
  issue_id?: number
  agent_task_id?: number
  extra?: Record<string, unknown>
}

export interface BuildListFilter {
  config_id?: number
  status?: BuildStatus
  project_id?: number
  limit?: number
}

export const cicdConfigApi = {
  list(workspaceId: number, filter: CICDConfigListFilter = {}): Promise<CICDConfig[]> {
    const params: Record<string, number> = {}
    if (filter.project_id) params.project_id = filter.project_id
    return api
      .get(`/workspaces/${workspaceId}/cicd/configs`, { params })
      .then(res => res.data)
  },

  get(workspaceId: number, configId: number): Promise<CICDConfig> {
    return api
      .get(`/workspaces/${workspaceId}/cicd/configs/${configId}`)
      .then(res => res.data)
  },

  create(workspaceId: number, payload: CICDConfigCreate): Promise<CICDConfig> {
    return api
      .post(`/workspaces/${workspaceId}/cicd/configs`, payload)
      .then(res => res.data)
  },

  update(workspaceId: number, configId: number, payload: CICDConfigUpdate): Promise<CICDConfig> {
    return api
      .patch(`/workspaces/${workspaceId}/cicd/configs/${configId}`, payload)
      .then(res => res.data)
  },

  delete(workspaceId: number, configId: number): Promise<void> {
    return api
      .delete(`/workspaces/${workspaceId}/cicd/configs/${configId}`)
      .then(() => undefined)
  }
}

export const cicdBuildApi = {
  list(workspaceId: number, filter: BuildListFilter = {}): Promise<BuildRecord[]> {
    const params: Record<string, string | number> = {}
    if (filter.config_id) params.config_id = filter.config_id
    if (filter.status) params.status = filter.status
    if (filter.project_id) params.project_id = filter.project_id
    if (filter.limit) params.limit = filter.limit
    return api
      .get(`/workspaces/${workspaceId}/cicd/builds`, { params })
      .then(res => res.data)
  },

  get(workspaceId: number, buildId: number): Promise<BuildRecord> {
    return api
      .get(`/workspaces/${workspaceId}/cicd/builds/${buildId}`)
      .then(res => res.data)
  },

  trigger(workspaceId: number, payload: BuildTriggerRequest): Promise<BuildRecord> {
    return api
      .post(`/workspaces/${workspaceId}/cicd/builds`, payload)
      .then(res => res.data)
  },

  cancel(workspaceId: number, buildId: number): Promise<BuildRecord> {
    return api
      .post(`/workspaces/${workspaceId}/cicd/builds/${buildId}/cancel`)
      .then(res => res.data)
  },

  delete(workspaceId: number, buildId: number): Promise<void> {
    return api
      .delete(`/workspaces/${workspaceId}/cicd/builds/${buildId}`)
      .then(() => undefined)
  }
}
